### ISSUE TITLE
Client-Side Signing

### ISSUE THREAD

**Issue by tresf** [qzind/tray#1]
Provide a code example for client-side signing:

Our guide offers many server-side signing examples:
 * https://github.com/qzind/qz-print/wiki/Signing-Messages

But no client-side examples.  The only method I'm aware of is through JavaScript which has the security risk of exposing the private key to all the users.

Our [Node.js example `sign-message.js`](https://github.com/qzind/qz-print/blob/1.9/assets/signing/sign-message.js) is a good place to start.

**Comment by tresf** [comment]
Migrated from https://github.com/qzind/qz-print/issues/99

***

@augugiacoEnta wrote:
Hi everyone,

If you need sign messages on the client-side, you can take this approach:

`var publicKey = "your public key";`

```
 qz.security.setCertificatePromise(function (resolve, reject) {
            //$.ajax("/assets/secure/qz/digital-certificate.txt").then(resolve, reject);
            resolve(publicKey);
        });
```


   ```
qz.security.setSignaturePromise(function (toSign) {
            return function (resolve, reject) {
	
		var privateKey = "your private key";
                var certificate509 = new X509();
                certificate509.readCertPEM(publicKey);
                var sha1Encryptator = certificate509.subjectPublicKeyRSA;
                sha1Encryptator.readPrivateKeyFromPEMString(privateKey);
                var messageSignedHex = sha1Encryptator.signString(toSign, 'sha1');
                var messageSignedStr = hextorstr(messageSignedHex);
                var finallyMsgSignedBase64 = stob64(messageSignedStr);

                resolve(finallyMsgSignedBase64);

            };
        });	
```

I used an external library called **jsrsasign**
Take a look here: http://kjur.github.io/jsrsasign/

It worked fine for me and my app is more fast that before.

I thought this info would be util for you.

Regards.

**Comment by tresf** [comment]
@augugiacoEnta, I'm trying out your example with one of our generated keys... 

```
sample.html:825 key is too short for SigAlg: keylen=246,sha1
```

So I changed our portal to generate a 512-bit key and then I get:
```
sample.html:770 RSASetPrivateEx N.length == 0
```

If I remove the `-----BEGIN PRIVATE KEY...` in the PEM I get:
```
sample.html:770 key is too short for SigAlg: keylen=99,sha1
```

I've tried loading the private key using your mechanism as well as 
```
// your technique
new X509().readPrivateKeyFromPEMString(privateKey)

// alternative technique
new RSAKey(privateKey);
```

Did you have to do anything special with your private key to get it to work with `jsrsasign`?

**Comment by augugiacoEnta** [comment]
Tres,

Yes,my apologies for this mistake.

I have to convert my private key in the new format to the old format with open-ssl for this works ok with jsrsasing.

Try this please:

`openssl rsa -in youPrivateKey.key -out yourPrivateKeyOldFormat.key`

Regards.

**Comment by tresf** [comment]
@augugiacoEnta

Much thanks!

Here's a minimalist working example:

```html
<script type="text/javascript" src="js/jsrsasign-latest-all-min.js"></script>
```

```javascript
qz.security.setSignaturePromise(function(toSign) {
   return function(resolve, reject) {
      var pk = new RSAKey(); pk.readPrivateKeyFromPEMString(privateKey);
      var hex = pk.signString(toSign, 'sha1');
      resolve(stob64(hextorstr(hex)));
   };		
});
```

**Comment by tresf** [comment]
Added to examples here:
https://github.com/qzind/tray/blob/2.0/assets/signing/sign-message.js

Closed via aa4207a

**Comment by bratpeki** [comment]
Accidentally named my commit "Fix # 1" in reference to a previous commit and ended up referencing this, lol
